src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -362,30 +398,45 @@ def infer_expr(
     - domain: The domain `expr` is read at.
     - symbolic_domain_sizes: A dictionary mapping axes names, e.g., `I`, `Vertex`, to a symbol
       name that evaluates to the length of that axis.
+    - allowed_unknown_domains: A list of references (as strings) for which the domain does not need
+      to be inferred, but can be `UNKNOWN`. This is used by `infer_program` to allow dynamic shifts
+      on program inputs.
 
     Returns:
       A tuple containing the inferred expression with all applied `as_fieldop` (that are accessed)
       having a domain argument now, and a dictionary mapping symbol names referenced in `expr` to
       domain they are accessed at.
     """
-    # this is just a small wrapper that populates the `domain` annex
-    expr, accessed_domains = _infer_expr(expr, domain, offset_provider, symbolic_domain_sizes)
+    allowed_unknown_domains = allowed_unknown_domains or []
+
+    expr, accessed_domains = _infer_expr(
+        expr, domain, offset_provider, symbolic_domain_sizes, allowed_unknown_domains
+    )
     expr.annex.domain = domain
+
+    if any(
+        p not in allowed_unknown_domains and d == "UNKNOWN" for p, d in accessed_domains.items()
+    ):
+        raise ValueError("Some accessed domains are unknown, e.g. because of a dynamic shift.")
+
     return expr, accessed_domains
 
 
 def _infer_stmt(
     stmt: itir.Stmt,
     offset_provider: common.OffsetProvider,
     symbolic_domain_sizes: Optional[dict[str, str]],
+    allowed_unknown_domains: list[str],
```

Comment:
In general, I am confused why this is introduced... 
Are all program parameters allowed to be unknown, but the rest isn't?

Reply:
If I understand the logic, then I would remove it from here and rather write a separate path for checking this and avoid uglifying the this interface.